header-rewriter/SourceRewriter.cpp

```diff
@@ -0,0 +1,632 @@
+#include "DetermineAbi.h"
+#include "GenCallAsm.h"
+#include "TypeOps.h"
+#include "clang/AST/AST.h"
+#include "clang/AST/Type.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/Basic/FileManager.h"
+#include "clang/Basic/SourceManager.h"
+#include "clang/Frontend/FrontendActions.h"
+#include "clang/Tooling/CommonOptionsParser.h"
+#include "clang/Tooling/Refactoring.h"
+#include "clang/Tooling/RefactoringCallbacks.h"
+#include "clang/Tooling/Tooling.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/Host.h"
+#include "llvm/Support/raw_ostream.h"
+#include <fstream>
+#include <optional>
+
+using namespace clang::ast_matchers;
+using namespace clang::tooling;
+using namespace std::string_literals;
+
+// Apply a custom category to all command-line options so that they are the
+// only ones displayed.
+static llvm::cl::OptionCategory
+    SourceRewriterCategory("Source rewriter options");
+
+static llvm::cl::opt<std::string>
+    OutputFilename("output-filename", llvm::cl::Required,
+                   llvm::cl::cat(SourceRewriterCategory),
+                   llvm::cl::desc("<prefix for output files>"));
+
+// Specifies the compartment's protection key index, if any.
+static llvm::cl::opt<uint32_t>
+    CompartmentPkey("compartment-pkey", llvm::cl::Required,
+                    llvm::cl::desc("The compartment's protection key index"),
+                    llvm::cl::cat(SourceRewriterCategory));
+
+static std::string get_filename(const clang::Decl *decl) {
+  assert(decl != nullptr);
+  const auto &sm = decl->getASTContext().getSourceManager();
+  return sm.getFilename(sm.getSpellingLoc(decl->getLocation())).str();
+}
+
+static std::string get_expansion_filename(const clang::Decl *decl) {
+  assert(decl != nullptr);
+  const auto &sm = decl->getASTContext().getSourceManager();
+  return sm.getFilename(sm.getExpansionLoc(decl->getLocation())).str();
+}
+
+static bool ignore_file(const std::string &file_name) {
+  auto file = llvm::StringRef(file_name);
+  // TODO: Add a proper blacklist with libia2 headers in it by default
+  return file.startswith("/usr/") || file.endswith("ia2.h") ||
+         file.endswith("test_fault_handler.h");
+}
+
+static std::string append_name_if_nonempty(const std::string &new_type,
+                                           const std::string &name) {
+  return new_type + (name.empty() ? "" : " ") + name;
+};
+
+struct FnPtrTy {
+  std::string fn_ptr;
+  std::string opaque_struct;
+
+  bool operator==(const FnPtrTy &other) const {
+    return (this->fn_ptr == other.fn_ptr) &&
+           (this->opaque_struct == other.opaque_struct);
+  };
+  bool operator<(const FnPtrTy &other) const {
```

Comment:
Is it possible for `this->fn_ptr < other.fn_ptr` and `this->opaque_struct > other.opaque_struct` to both be true? Then `==` and `<` will always return `false`.

Reply:
Removed this struct.